lib/Twig/Loader/Array.php

```diff
@@ -80,7 +80,7 @@ public function getCacheKey($name)
             throw new Twig_Error_Loader(sprintf('Template "%s" is not defined.', $name));
         }
 
-        return $this->templates[$name];
+        return $name.$this->templates[$name];
```

Comment:
@stof Dangit, I thought about that, but `\Twig_Environment::getTemplateClass` doesn't make this distinction when it appends the optionsHash (https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Environment.php#L269), so I skipped it.

It's definitely possible to get a collision in the method I provided though; I'll add something in.